### PR TITLE
Laboratory work #3, Kramkova Maria - 22FPL1

### PR DESCRIPTION
**Pull request by kramkova**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
good. But change score

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +43,18 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not (isinstance(text, str)) or not text:
+            return None
+        preprocessed = []
+        for symbol in text.lower():
+            if symbol.isalpha():
+                preprocessed.append(symbol)
+            if (symbol in '!"#$%&\'()*+, ./:;<=>?@[]^_`{|}~\\'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:52]: `string.punctuation`

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -76,6 +94,10 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+        position = list(self._storage.values()).index(element_id)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:99]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -93,6 +115,18 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
+        if not isinstance(text, str):
+            return None
+        unigrams = self._tokenize(text)
+        if not unigrams:
+            return None
+        encoded = []
+        for symbol in unigrams:
+            self._put(symbol)
+            encoded.append(self.get_id(symbol))
+        if None in self._storage or None in encoded:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:127]: check it in for

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -104,6 +138,10 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:143]: erly return

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (kramkova) [reply lab_3_generate_by_ngrams/main.py:52]: done

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (kramkova) [reply lab_3_generate_by_ngrams/main.py:127]: done

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (kramkova) [reply lab_3_generate_by_ngrams/main.py:143]: attempted

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (kramkova) [reply lab_3_generate_by_ngrams/main.py:99]: implemented!

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:129]: are you sure?

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -143,6 +188,14 @@ def _decode(self, corpus: tuple[int, ...]) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(corpus, tuple) or not corpus:
+            return None
+        decoded = []
+        for token_id in corpus:
+            decoded.append(self.get_token(token_id))
+        if None in decoded:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:196]: move to the loop. and return as soon as posible?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good implementation but please, add some changes

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (kramkova) [reply lab_3_generate_by_ngrams/main.py:196]: attempted

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (kramkova) [reply lab_3_generate_by_ngrams/main.py]: sorry! it has changed now

**Pull request closed without merging**